src/Build/Logging/BaseConsoleLogger.cs

```diff
@@ -911,6 +891,195 @@ public int Compare(object o1, object o2)
+                            pOutput += segment.Length;
+
+                            // append newLine
+                            for (int i = 0; i < newLine.Length; i++)
+                            {
+                                *pOutput++ = newLine[i];
+                            }
+
+                            // move to next segment
+                            pSegment += segment.TotalLength;
+                        }
+                    }
+#pragma warning restore SA1519 // Braces should not be omitted from multi-line child statement
+                }
+
+                return indented;
+
+                // local method
+                static int ComputeIndentedStringLength(PooledSpan<StringSegment> segments, string newLine, int indent)
+                {
+                    int indentedLength = segments.Length * (newLine.Length + indent);
+
+                    foreach (var segment in segments)
+                    {
+                        indentedLength += segment.Length;
+                    }
+
+                    return indentedLength;
+                }
+            }
+
+            private static PooledSpan<StringSegment> GetStringSegments(string input)
+            {
+                if (input.Length == 0)
+                {
+                    PooledSpan<StringSegment> emptyResult = new(1);
+                    emptyResult.Span[0] = new StringSegment(0, 0);
+                    return emptyResult;
+                }
+
+                int segmentCount = 1;
+                for (int i = 0; i < input.Length; i++)
+                {
+                    if (input[i] == '\n')
+                    {
+                        segmentCount++;
+                    }
+                }
+
+                PooledSpan<StringSegment> segments = new(segmentCount);
+                int start = 0;
+                int index = 0;
+
+                for (int i = 0; i < segmentCount; i++)
+                {
+                    while (index < input.Length && input[index] != '\n')
+                    {
+                        index++;
+                    }
+
+                    // the input string didn't end with a newline
+                    if (index == input.Length)
+                    {
+                        segments[i] = new StringSegment(index - start, 0);
+                        break;
+                    }
+
+                    int newLineLength = 1;
+                    bool endedWithReturnNewline = (index > 0) && (input[index - 1] == '\r');
+
+                    if (endedWithReturnNewline)
+                    {
+                        newLineLength++;
+                        index--;
+                    }
+
+                    segments[i] = new StringSegment(index - start, newLineLength);
+
+                    start = index += newLineLength;
+                }
+
+                return segments;
+            }
+
+            private readonly record struct StringSegment(int Length, int NewLineLength)
```

Comment:
Do you need all the extra code that record generates?

Reply:
Yowza!  TIL...

Reply:
Now using `class` instead of `record`

Reply:
This becomes an allocation if you use class.

Reply:
My bad.  We are [now using](https://github.com/dotnet/msbuild/pull/8934/files#diff-2ff34608fe924cf84311adfb6f05a9ccb2a73b4aec7e98c67be78436a4ff20a3:~:text=private%20struct%20StringSegment) `struct` instead of `record struct`.